[State Mgmt] Docset refresh (#2541)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-get-save-state.md

description: "Use key value pairs to persist a state"
---

> State management is one of the most common needs of any new, legacy, monolith, or microservice application. Dealing with and testing different database libraries and handling retries and faults can be both diffictult and time consuming.

Reviewer: ```suggestion
State management is one of the most common needs of any new, legacy, monolith, or microservice application. Dealing with and testing different database libraries and handling retries and faults can be both difficult and time consuming.
```

In this guide, you'll learn the basics of using the key/value state API to allow an application to save, get, and delete state.

## Example

> The below code example _loosely_ describes an application that processes orders with an order processing service which has a Dapr sidecar. The order processing service uses Dapr to store state in a Redis state store.

Reviewer: ```suggestion
The code example below _loosely_ describes an application that processes orders with an order processing service which has a Dapr sidecar. The order processing service uses Dapr to store state in a Redis state store.
```

{{< tabs "Self-Hosted (CLI)" Kubernetes>}}

{{% codetab %}}

> When you run `dapr init` in standalone mode, Dapr creates a default Redis `statestore.yaml` and runs a Redis state store on your local machine, located:

Reviewer: ```suggestion
When you run `dapr init` in self hosted mode, Dapr creates a default Redis `statestore.yaml` and runs a Redis state store on your local machine, located:
```

- On Windows, under `%UserProfile%\.dapr\components\statestore.yaml`
- On Linux/MacOS, under `~/.dapr/components/statestore.yaml`

With the `statestore.yaml` component, you can easily swap out underlying components without application code changes.

{{% /codetab %}}

{{% codetab %}}

> To deploy this into a Kubernetes cluster, fill in the `metadata` connection details of your [desired state store component]({{< ref supported-state-stores >}}) in the yaml below, save as `statestore.yaml`, and run `kubectl apply -f statestore.yaml`.

Reviewer: ```suggestion
To deploy this into a Kubernetes cluster, fill in the `metadata` connection details of your [state store component]({{< ref supported-state-stores >}}) in the YAML below, save as `statestore.yaml`, and run `kubectl apply -f statestore.yaml`.
```

  - name: redisPassword
    value: ""
```

See [how to setup different state stores on Kubernetes]({{< ref "setup-state-store" >}}).

{{% /codetab %}}

{{< /tabs >}}

{{% alert title="Important" color="warning" %}}

> Set an app-id, as the state keys are prefixed with this value. If you don't set an app-id, one is generated for you at runtime. The next time you run the command, a new app-id will be generated and you will no longer have access to the previously saved state.

Reviewer: ```suggestion
Set an `app-id`, as the state keys are prefixed with this value. If you don't set an `app-id`, one is generated for you at runtime. The next time you run the command, a new `app-id` is generated and you will no longer have access to the previously saved state.
```

{{% /alert %}}

## Save and retrieve a single state

> The following example shows how to save and retrieve a single key/value pair using the Dapr state building block.

Reviewer: ```suggestion
The following example shows how to save and retrieve a single key/value pair using the Dapr state management API.
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

        logging.info('Result after get: ' + result.data.decode('utf-8'))
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

main();
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

    client.delete_state(store_name=DAPR_STORE_NAME, key="order_1")
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

main();
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

    logging.info('Result after get bulk: ' + str(result)) 
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

main();
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

{{< /tabs >}}

## Perform state transactions

{{% alert title="Note" color="primary" %}}

> State transactions require a state store that supports multi-item transactions. Visit the [supported state stores page]({{< ref supported-state-stores >}}) for a full list.

Reviewer: ```suggestion
State transactions require a state store that supports multi-item transactions. See the [supported state stores page]({{< ref supported-state-stores >}}) for a full list.
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

}
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

    logging.info('Result: ' + str(result))
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```

main();
```

> To launch a Dapr sidecar and run the above example application, you'd then run a command similar to the following:

Reviewer: ```suggestion
To launch a Dapr sidecar for the above example application, run a command similar to the following:
```
